server/routerlicious/packages/services-shared/src/socketIoServer.ts

```diff
@@ -127,13 +128,61 @@ class SocketIoServer implements core.IWebSocketServer {
 	}
 
 	public async close(): Promise<void> {
+		if (this.socketIoConfig?.gradualDisconnectEnabled) {
+			// Gradual disconnection
+			const drainTime = this.socketIoConfig?.gradualDisconnectDrainTimeMs ?? 30000;
+			const drainInterval = this.socketIoConfig?.gradualDisconnectDrainIntervalMs ?? 1000;
+			if (drainTime > 0 && drainInterval > 0) {
+				// we are assuming no new connections appear once we start. any leftover connections will be closed when close is called
+				const connections = await this.io.fetchSockets();
+				const connectionCount = connections.length;
+				Lumberjack.info("Gradual disconnection started", {
+					drainTime,
+					drainInterval,
+					connectionCount,
+				});
+				// total number of drains to run
+				const totalDrains = Math.ceil(drainTime / drainInterval);
+				// number of connections to disconnect per drain
+				const connectionsToDisconnectPerDrain = Math.ceil(connectionCount / totalDrains);
+				let done = false;
+				const drainConnections = Array.from(connections.values());
+				if (connectionsToDisconnectPerDrain > 0) {
+					// start draining                let done = false;
+					let n = 0;
+					for (let i = 0; i < totalDrains; i++) {
```

Comment:
Should we put this in try/catch/finally block?
What happens when this code throw exception?

Reply:
with that let's add a metric for this flow. It will be easier to alert on it.